python/ucxx/ucxx/_lib_async/continuous_ucx_progress.py

```diff
@@ -70,3 +77,133 @@ async def _progress_task(self):
+        # return when all non-IO tasks are finished.
+        # See <https://stackoverflow.com/a/48491563>.
+        self.rsock, wsock = socket.socketpair()
+        self.rsock.setblocking(0)
+        wsock.setblocking(0)
+        wsock.close()
+
+        epoll_fd = worker.epoll_file_descriptor
+
+        # Bind an asyncio reader to a UCX epoll file descriptor
+        event_loop.add_reader(epoll_fd, self._fd_reader_callback)
+
+        # Remove the reader and close socket on finalization
+        weakref.finalize(self, event_loop.remove_reader, epoll_fd)
+        weakref.finalize(self, self.rsock.close)
+
+        self.blocking_asyncio_task = None
+        self.last_progress_time = time.monotonic() - progress_timeout
+        self.asyncio_task = event_loop.create_task(self._timeout_progress(1.0))
+
+    def __del__(self):
+        """Cancel asynchronous blocking progress task.
+
+        Cancel asynchronouns blocking progress task.
+
+        .. warning::
+            This only works if the event loop is still running. If the event loop has
+            been closed before this runs the following error will be printed by the
+            interpreter on the standard output:
+
+            ```
+            Task was destroyed but it is pending!
+            ```
+        """
+        if self.event_loop is not None and self.event_loop.is_running():
+            if self.blocking_asyncio_task is not None:
+                self.call_soon_threadsafe(self.blocking_asyncio_task.cancel())
+
+        super().__del__()
+
+    def _fd_reader_callback(self):
+        """Schedule new progress task upon worker event.
+
+        Schedule new progress task when a new event occurs in the worker's epoll file
+        descriptor.
+        """
+        self.worker.progress()
+
+        # Notice, we can safely overwrite `self.dangling_arm_task`
+        # since previous arm task is finished by now.
+        assert self.blocking_asyncio_task is None or self.blocking_asyncio_task.done()
+        self.blocking_asyncio_task = self.event_loop.create_task(self._arm_worker())
+
+    async def _arm_worker(self):
+        """Progress the worker and rearm.
+
+        Progress and rearm the worker to watch for new events on its epoll file
+        descriptor.
+        """
+        # When arming the worker, the following must be true:
+        #  - No more progress in UCX (see doc of ucp_worker_arm())
+        #  - All asyncio tasks that isn't waiting on UCX must be executed
+        #    so that the asyncio's next state is epoll wait.
+        #    See <https://github.com/rapidsai/ucx-py/issues/413>
+        while True:
+            self.last_progress_time = time.monotonic()
+            self.worker.progress()
+
+            # This IO task returns when all non-IO tasks are finished.
+            # Notice, we do NOT hold a reference to `worker` while waiting.
+            await self.event_loop.sock_recv(self.rsock, 1)
```

Comment:
**Problem**: Potential deadlock if `self.event_loop.sock_recv(self.rsock, 1)` blocks indefinitely.

**Suggested Change**: Use `asyncio.wait_for` to add a timeout to the `sock_recv` call.


**Severity (1 - 4)**: 4 - ***CRITICAL***

**Line**: 179

```suggestion
            ```python
            await asyncio.wait_for(self.event_loop.sock_recv(self.rsock, 1), timeout
```

-----

<div align="right">
    <small><i>Generated by <b>Llama 3.1 405b</b></i></small>
</div>
<br />


<div align="right">
  <small>Was this helpful?</small>
  <a href="https://codecritic-feedback.nvidia.com?vote=1&comment_guid=ac3f23fc-7575-45e8-a832-764806208fd4&is_github=true" target="_blank">
    <button>👍</button>
  </a>
  <a href="https://codecritic-feedback.nvidia.com?vote=-1&comment_guid=ac3f23fc-7575-45e8-a832-764806208fd4&is_github=true" target="_blank">
    <button>👎</button>
  </a>
</div>